### PR TITLE
Migrate redhat importer

### PR DESCRIPTION
**Pull request by TG1999**
Signed-off-by: Tushar Goel <tushar.goel.dav@gmail.com>

### CODE REVIEW

File: vulnerabilities/importers/redhat.py [hunk 1]
```diff
@@ -20,73 +20,83 @@
...
         if not cve_data:
             break
-        cves.extend(cve_data)
         page_no += 1
+        yield cve_data
+
+
+class RedhatImporter(Importer):
+
+    spdx_license_expression = "CC-BY-4.0"
```
Comment (pombredanne) [inline vulnerabilities/importers/redhat.py:71]: Is there a license URL too?

File: vulnerabilities/importers/redhat.py [hunk 2]
```diff
@@ -20,73 +20,83 @@
...
-    affected_purls = []
-    if advisory_data.get("affected_packages"):
-        for rpm in advisory_data["affected_packages"]:
-            purl = rpm_to_purl(rpm)
-            if purl:
-                affected_purls.append(purl)
+    affected_packages: List[AffectedPackage] = []
+    for rpm in advisory_data.get("affected_packages") or []:
+        purl = rpm_to_purl(rpm)
+        if purl:
```
Comment (pombredanne) [inline vulnerabilities/importers/redhat.py:82]: What if there is no purl? are we skipping data silently then?

File: vulnerabilities/importers/redhat.py [hunk 3]
```diff
@@ -166,25 +179,98 @@ def to_advisory(advisory_data):
 
     references.append(Reference(severities=redhat_scores, url=advisory_data["resource_url"]))
     return AdvisoryData(
-        vulnerability_id=advisory_data["CVE"],
-        summary=advisory_data["bugzilla_description"],
-        affected_packages=nearest_patched_package(affected_purls, []),
+        aliases=advisory_data.get("CVE") or "",
```
Comment (pombredanne) [inline vulnerabilities/importers/redhat.py:182]: I think aliases is a list not a string for not a `or ""``

File: vulnerabilities/importers/redhat.py [hunk 4]
```diff
@@ -166,25 +179,98 @@ def to_advisory(advisory_data):
...
+    n, v, r, a = m.groups()
+    if ":" not in v:
+        return None, n, v, r, a
+    e, v = v.split(":", 1)
+    if e.isdigit():
+        e = int(e)
+    return (e, n, v, r, a)
+
+
 def rpm_to_purl(rpm_string):
```
Comment (pombredanne) [inline vulnerabilities/importers/redhat.py:247]: IMHO move this to an rpm_utils.py module too, and accept a new argument which would be the "namespace" (here redhat)

**Review by pombredanne — CHANGES_REQUESTED** [review]
Thanks! I posted a few review comments,
The key thing is to make sure we do not skip over data silently without logging.

File: vulnerabilities/importers/redhat.py (hunk 2 above)
Comment (TG1999) [reply vulnerabilities/importers/redhat.py:82]: We are not skipping the data silently, whenever we return None from rpm_to_purl function we log the reason why we can't return a purl, and since there could by many cases for returning a Purl none, we can't infer it at the calling side

File: vulnerabilities/importers/redhat.py [hunk 5]
```diff
@@ -20,73 +20,84 @@
...
     page_no = 1
-    url_template = "https://access.redhat.com/hydra/rest/securitydata/cve.json?per_page=10000&page={}"  # nopep8
-
     cve_data = None
     while True:
-        current_url = url_template.format(page_no)
+        current_url = f"https://access.redhat.com/hydra/rest/securitydata/cve.json?per_page=10000&page={page_no}"  # nopep8
         try:
-            print(f"Fetching: {current_url}")
             response = requests_session.get(current_url)
```
Comment (pombredanne) [inline vulnerabilities/importers/redhat.py:54]: IMHO the whole fetching should be abstracted in a function that hides requests/sessions/statsus code and processes all errors. This can be in another PR though

File: vulnerabilities/importers/redhat.py [hunk 6]
```diff
@@ -126,7 +141,7 @@ def to_advisory(advisory_data):
             rhsa_aggregate_severities = []
             if rhsa_data.get("cvrfdoc"):
                 # not all RHSA errata have a corresponding CVRF document
-                value = rhsa_data["cvrfdoc"]["aggregate_severity"]
+                value = get_item(rhsa_data, "cvrfdoc", "aggregate_severity")
```
Comment (pombredanne) [inline vulnerabilities/importers/redhat.py:144]: Let's handle the cases where there is no value
```suggestion
                value = get_item(rhsa_data, "cvrfdoc", "aggregate_severity")
                if value:
```

File: vulnerabilities/importers/redhat.py [hunk 7]
```diff
@@ -126,7 +141,7 @@ def to_advisory(advisory_data):
...
                 rhsa_aggregate_severities.append(
```
Comment (pombredanne) [inline vulnerabilities/importers/redhat.py:145]:
```suggestion
                    rhsa_aggregate_severities.append(
```

**Review by pombredanne — APPROVED** [review]
LGTM... Just a few minor comments

**Comment by TG1999** [comment]
Thanks @pombredanne , while I was at it, I realised severities were not displayed on UI, also fixed that. Merging it as soon it's all green!
